test/unit/org/apache/cassandra/tools/ToolRunner.java

```diff
@@ -176,17 +176,51 @@ public OutputStream getOutputStream()
                     }
 
                     @Override
-                    public int waitFor() throws InterruptedException
+                    public int waitFor()
                     {
                         return exitValue();
                     }
                     
                 };
             }
-            
-            ioWatcher = new Thread(this::watchIO);
-            ioWatcher.setDaemon(true);
-            ioWatcher.start();
+
+            // each stream tends to use a bounded buffer, so need to process each stream in its own thread else we
+            // might block on an idle stream, not consuming the other stream which is blocked in the other process
+            // as nothing is consuming
+            int numWatchers = 2;
+            // only need a stdin watcher when forking
+            boolean includeStdinWatcher = runOutOfProcess && stdin != null;
+            if (includeStdinWatcher)
+                numWatchers = 3;
+            ioWatchers = new Thread[numWatchers];
+            ioWatchers[0] = new Thread(new StreamGobbler<>(process.getErrorStream(), err));
+            ioWatchers[0].setDaemon(true);
+            ioWatchers[0].setName("IO Watcher stderr for " + allArgs);
+            ioWatchers[0].start();
+
+            ioWatchers[1] = new Thread(new StreamGobbler<>(process.getInputStream(), out));
+            ioWatchers[1].setDaemon(true);
+            ioWatchers[1].setName("IO Watcher stdout for " + allArgs);
+            ioWatchers[1].start();
+
+            if (includeStdinWatcher)
+            {
+                ioWatchers[2] = new Thread(new StreamGobbler<>(stdin, process.getOutputStream()));
+                ioWatchers[2].setDaemon(true);
+                ioWatchers[2].setName("IO Watcher stdin for " + allArgs);
+                ioWatchers[2].start();
+                // since stdin might not close the thread would block, so add logic to try to close stdin when the process exits
+                onComplete.whenComplete((i1, i2) -> {
+                    try
+                    {
+                        stdin.close();
```

Comment:
sadly this is leaky as the life-cycle isn't clearly monitored so there are edge cases where this may start to leak...